src/type/validate.js

```diff
@@ -261,6 +267,20 @@ function validateTypes(context: SchemaValidationContext): void {
   });
 }
 
+function validateScalarSerialization(
+  context: SchemaValidationContext,
+  scalarType: GraphQLScalarType,
+): void {
+  if (scalarType.ofType && !isSpecifiedScalarType(scalarType.ofType)) {
```

Comment:
I wonder if this is over-constrained. Perhaps it would be okay to allow a scalar to serialize in terms of any other scalar type?

My thought is that people add custom scalars for two reasons. First is to represent some higher level concept like "Email" or "PhoneNumber" where "serializes as String" is useful. Second is to represent a new kind of transport primitive like "BigInt" where they definitionally serialize as something new. However that first kind of higher-level concept scalar might want to say it serializes in terms of another non-spec scalar!

In other words - what makes the specification of a scalar special to warrant this limitation?

I believe we originally settled on this limitation to avoid needing to follow a chain of "ofType" to determine the final actual serialization type, but I think that might not be such a big deal

Reply:
> I believe we originally settled on this limitation to avoid needing to follow a chain of "ofType" to determine the final actual serialization type, but I think that might not be such a big deal

I think that decision about allowing recursive type should be done in the scope of the entire spec since it affects the entire type system. For example, you couldn't query all information about type using `__type` anymore:
```graphql
{
   __type(name: "LocalPath") {
      name
      # ...
      ofType {
         name
         # ...
         # <= you can't construct a query that gets arbitrary long chain of `ofType`
      }
   }
}
```
Or the fact that you can introduce loops inside types, so we need to make `buildASTSchema`, `buildClientSchema` and others loop-resistant. E.g.:
```graphql
scalar A as C
scalar B as A
scalar C as B
```

I think allowing to better describe scalars isn't worth added complexity. However, if the similar changes are done in other places of type-system (e.g. https://github.com/facebook/graphql/issues/295) then the price is already paid and we could definitely add recursiveness to scalar types.

> However that first kind of higher-level concept scalar might want to say it serializes in terms of another non-spec scalar!

We can allow specifying scalar serializable as any other scalar without `as` in the definition.
```graphql
scalar BigInt
scalar Counter as BigInt # Valid
scalar UserCounter as Counter # Invalid
```
But with current terminology and SDL syntax, it would be very confusing 😞 
That's why I think we need to separate `custom scalar type`s and `scalar type`s:
```graphql
custom scalar BigInt
scalar Counter as BigInt # Valid
scalar UserCounter as Counter # Invalid
```
So we could say that scalar is serializable only as a standard scalar or a custom scalar.

-------------------------------
I think this PR in its current form solves >80% of use-cases so we can release it as is and relax rules afterward.